lib/SILOptimizer/SILCombiner/SILCombinerMiscVisitors.cpp

```diff
@@ -1761,8 +1761,13 @@ visitAllocRefDynamicInst(AllocRefDynamicInst *ARDI) {
   Builder.setCurrentDebugScope(ARDI->getDebugScope());
 
   SILValue MDVal = ARDI->getMetatypeOperand();
-  while (auto *UCI = dyn_cast<UpcastInst>(MDVal))
+  while (auto *UCI = dyn_cast<UpcastInst>(MDVal)) {
+    // For simplicity ignore a cast of an `alloc_ref [stack]`. It would need more
+    // work to keep its `dealloc_stack_ref` correct.
+    if (ARDI->canAllocOnStack())
```

Comment:
Would it make sense to just check this condition once at the beginning of the function rather than in a loop here and then again below in the `auto *AI = dyn_cast<ApplyInst>(MDVal)` case?